operator/builtin/input/aws/cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,372 @@
+			Limit:          limit,
+			LogGroupName:   logGroupNamePtr,
+			LogStreamNames: c.logStreamNames,
+			StartTime:      startTime,
+		}
+	}
+
+	if nextToken != "" {
+		nextTokenPtr := aws.String(nextToken)
+		return cloudwatchlogs.FilterLogEventsInput{
+			Limit:        limit,
+			LogGroupName: logGroupNamePtr,
+			StartTime:    startTime,
+			NextToken:    nextTokenPtr,
+		}
+	}
+
+	return cloudwatchlogs.FilterLogEventsInput{
+		Limit:        limit,
+		LogGroupName: logGroupNamePtr,
+		StartTime:    startTime,
+	}
+}
+
+// handleEvent is the handler for a AWS Cloudwatch Logs Filtered Event.
+func (c *CloudwatchInput) handleEvent(ctx context.Context, event *cloudwatchlogs.FilteredLogEvent) error {
+	e := make(map[string]interface{})
+	e["message"] = event.Message
+	e["event_id"] = event.EventId
+	e["log_stream_name"] = event.LogStreamName
+	e["ingestion_time"] = event.IngestionTime
+
+	entry, err := c.NewEntry(nil)
+	if err != nil {
+		return errors.Wrap(err, "Failed to create new entry from record")
+	}
+
+	entry.Timestamp = fromUnixMilli(*event.Timestamp)
+	entry.Record = e
+
+	// Persist
+	if *event.IngestionTime > c.startTime {
+		c.startTime = *event.IngestionTime
+		c.Debugf("Writing start time %d to database", *event.IngestionTime)
+		c.persist.Write(c.logGroupName, *event.IngestionTime)
+	}
+
+	// Write Entry
+	c.Write(ctx, entry)
+	return nil
+}
+
+func (c *CloudwatchInput) handleEvents(ctx context.Context, events []*cloudwatchlogs.FilteredLogEvent) error {
+	c.wg.Add(1)
+	defer c.wg.Done()
+
+	for _, event := range events {
+		c.handleEvent(ctx, event)
+	}
+	c.persist.DB.Sync()
+	return nil
+}
+
+// Returns time.Now() as Unix Time in Milliseconds
+func currentTimeInUnixMilliseconds() int64 {
+	return time.Now().UnixNano() / int64(time.Millisecond)
+}
+
+// Helper function to convert Unix epoch time in milliseconds to go time
+func fromUnixMilli(ms int64) time.Time {
+	const millisInSecond = 1000
+	const nsInSecond = 1000000
+	return time.Unix(ms/int64(millisInSecond), (ms%int64(millisInSecond))*int64(nsInSecond))
+}
+
+// timeLayoutParser parses set of predefined variables and replaces with date equivalent
+func (c *CloudwatchInput) timeLayoutParser(layout string) string {
+	if strings.Contains(layout, "%") {
+		replace := map[string]string{
+			"%Y": "2006",    // Year, zero-padded
+			"%y": "06",      // Year, last two digits, zero-padded
+			"%m": "01",      // Month as a decimal number
+			"%q": "1",       // Month as a unpadded number
+			"%b": "Jan",     // Abbreviated month name
+			"%h": "Jan",     // Abbreviated month name
+			"%B": "January", // Full month name
+			"%d": "02",      // Day of the month, zero-padded
+			"%g": "2",       // Day of the month, unpadded
+			"%a": "Mon",     // Abbreviated weekday name
+			"%A": "Monday",  // Full weekday name
+		}
+
+		for key, value := range replace {
+			layout = strings.Replace(layout, key, value, 1)
+		}
+		return time.Now().Format(layout)
+	}
+	return layout
+}
```

Comment:
I think we should test these. Generally we test the "public" interface, but in this case, the public interface is not being tested and it relies heavily on these 3 functions.